base/threadingconstructs.jl

```diff
@@ -175,9 +175,46 @@ end
 function _threadsfor(iter, lbody, schedule)
     lidx = iter.args[1]         # index
     range = iter.args[2]
+    esc_range = esc(range)
+    func = if schedule === :greedy
+        greedy_func(esc_range, lidx, lbody)
+    else
+        default_func(esc_range, lidx, lbody)
+    end
     quote
         local threadsfor_fun
-        let range = $(esc(range))
+        $func
+        if $(schedule === :greedy || schedule === :dynamic || schedule === :default)
+            threading_run(threadsfor_fun, false)
+        elseif ccall(:jl_in_threaded_region, Cint, ()) != 0 # :static
+            error("`@threads :static` cannot be used concurrently or nested")
+        else # :static
+            threading_run(threadsfor_fun, true)
+        end
+        nothing
+    end
+end
+
+function greedy_func(itr, lidx, lbody)
+    quote
+        let c = Channel{eltype($itr)}(threadpoolsize(),spawn=true) do ch
```

Comment:
It still seems odd for the channel to always contain 2x the work that it should have contained if the work was being produced only as fast as they were being worked on, as it also doesn't seem to accurately reflect the docstring:

> working on the given iterated values as they are produced

```suggestion
        let c = Channel{eltype($itr)}(0, spawn=true) do ch
```

Reply:
What do you mean with 2x? The channel only holds as much as the amount of tasks used to run. The individual worker tasks take from that channel, so using `Channel(0)` i.e. an unbuffered channel will IMO unnecessarily increase contention on the channel lock (though to be fair, I haven't analysed whether that is actually the case/leads to a performance loss).

"as they are produced" is supposed to be read as "in whatever order the iterator produces them", not as a guarantee that the work starts immediately after `iterate` returns.

Reply:
It holds an extra work item per task, so 2x the number of tasks get allocated (half of which are being worked on). How is this supposed to reduce lock contention? Now you are doing `threadpoolsize` extraneous operations on the Channel that weren't there with an unbuffered channel, each of which needs to take the lock.

Reply:
> It holds an extra work item per task, so 2x the number of tasks get allocated (half of which are being worked on).

What is being allocated though? Do you mean the array in the underlying channel? I don't see how you get to double the memory, instead of just 1x the memory. Objects that would be copied around either way shouldn't increase that, and objects that are stored as references also will at most occupy a pointer.

> How is this supposed to reduce lock contention? Now you are doing threadpoolsize extraneous operations on the Channel that weren't there with an unbuffered channel, each of which needs to take the lock.

I thought that the buffered channel was implemented as a circular buffer, where taking is more or less free if there are items available, and where putting is more or less free if there is space available (no data needs to be copied there and only the read/write heads need to be secured with `@atomic`). Looking at the implementation, this actually does a regular `popfirst!` & `push!` on the data array though, so the thinking doesn't apply because this might have to move pretty much all of that data around. A bit unfortunate, but I guess that means staying unbuffered will be better here.